### PR TITLE
Replace all internal usage of `get_upstream` with `get_upstream_resource`

### PR DESCRIPTION
**Pull request by miscco**
We want to get away from raw resources, so prepare deprecation of it by replacing all internal usages

This PR relies on preparation in downstream repositories

### CODE REVIEW

**Review by bdice — APPROVED** [review]

**Review by vyasr — APPROVED** [review]

**Review by harrism — CHANGES_REQUESTED** [review]
Requesting changes to bake sure this doesn't merge before the downstream PRs are merged.

**Comment by harrism** [comment]
@miscco if you would like to merge this before downstream PRs, you could move the deprecations to a separate PR that won't be merged yet.

**Comment by bdice** [comment]
I'd favor splitting this PR into the new functions and then the deprecations, it does seem to help with moving the work along.

**Comment by miscco** [comment]
As far as I can see, there are two occurrences of `get_upstream` outside of rmm

https://github.com/rapidsai/cudf/pull/15203
https://github.com/rapidsai/raft/pull/2207

The only one I am not 100% sure how to get rid of is one other in raft where it explicitly checks whether a resource of of a given type, which is neither possible with the current `resource_ref` not is it desired

Given that those PRs are very small and limited in scope I believe we can wait a few days getting them in

**Comment by harrism** [comment]
Also the RMM Cython needs to be updated, and all uses of its current method.

**Review by harrism — APPROVED** [review]

**Comment by harrism** [comment]
@miscco does this PR also result in changing the semantics of MR equality as discussed in #1402 ?

File: include/rmm/mr/device/binning_memory_resource.hpp [hunk 1]
```diff
@@ -151,11 +151,11 @@ class binning_memory_resource final : public device_memory_resource {
    * @param bytes Requested allocation size in bytes
    * @return rmm::mr::device_memory_resource& memory_resource that can allocate the requested size.
```
Comment (wence-) [inline include/rmm/mr/device/binning_memory_resource.hpp:152]: *nit*: `rmm::async_resource_ref`? And maybe change the `@brief` to "Get the resource reference for the requested size"?

**Review by wence- — COMMENTED** [review]
Tiny doc nitpick.

**Comment by wence-** [comment]
> Also the RMM Cython needs to be updated, and all uses of its current method.

We decided that this didn't seem to be necessary, since, for now, the Cython `get_upstream` function doesn't call into C++ (it returns a member variable of the cython class).

**Pull request merged**